lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php

```diff
@@ -1021,15 +1021,59 @@ public function prepareQueryOrNewObj(array $query, bool $isNewObj = false) : arr
 
             $preparedQueryElements = $this->prepareQueryElement((string) $key, $value, null, true, $isNewObj);
             foreach ($preparedQueryElements as [$preparedKey, $preparedValue]) {
-                $preparedQuery[$preparedKey] = is_array($preparedValue)
-                    ? array_map('\Doctrine\ODM\MongoDB\Types\Type::convertPHPToDatabaseValue', $preparedValue)
-                    : Type::convertPHPToDatabaseValue($preparedValue);
+                $preparedQuery[$preparedKey] = $this->convertToDatabaseValue($key, $preparedValue);
             }
         }
 
         return $preparedQuery;
     }
 
+    /**
+     * Converts a single value to its database representation based on the mapping type
+     *
+     * @param string $fieldName
+     * @param mixed  $value
+     *
+     * @return mixed
+     */
+    private function convertToDatabaseValue($fieldName, $value)
```

Comment:
For new code, please use type declarations where possible:

```suggestion
    private function convertToDatabaseValue(string $fieldName, $value)
```